### PR TITLE
Cargo - Improve various aspects

### LINKED ISSUE #9109: ACE Cargo actions don't show up under special circumstances with JIP
**Mods (complete and add to the following information):**
- **Arma 3:** `2.10` profiling
- **CBA:** `3.15.8`
- **ACE3:** `3.15.2`

**Description:**
If the ACE cargo status for an item is changed mid-mission (turn on and then turn off), a new JIP player won't see further changes of the ACE Cargo status of that item, see steps to reproduce. If the status for that item wasn't changed before the new player joins, the problem doesn't occur. If there are no JIP players (all players logged in since server start) the problem also doesn't occur.

**Steps to reproduce:**
1. Start Empty Mission (containing only a random object and 2 player slots) in dedicated mp **or hosted mp**
2. join player1 (or start mission with already joinded player1)
3. enable ace cargo for object via debug console on player1 ( `[cursorObject, 1] call ace_cargo_fnc_setSize;` )
4. disable ace cargo for object via debug console on player1 ( `[cursorObject, -1] call ace_cargo_fnc_setSize;` )
5. join player2
6. enable ace cargo for object via debug console on player1 ( `[cursorObject, 1] call ace_cargo_fnc_setSize;` )

**Expected behavior:**
The JIP player2 should see the ACE3 Cargo interactions that player1 had enabled (Cargo and/or Rename)

**Where did the issue occur?**
- Dedicated / Hosted MP

**Log Files:**
- [client RPT](https://gist.github.com/y0014984/0b5fa41875a63bf84b06c94f9af15256)

**Additional context:**
I think the variable `ace_cargo_setSize_jipID` set on the object prohibits the expected behavior. If I execute the following command (deleting the var) on player1 right before enabling ACE3 Cargo, the problem never occurs:
```sqf
cursorObject setVariable ["ace_cargo_setSize_jipID", nil, true];
```

**Screenshots:**
There are no helpful screenshots, except showing the missing interactions.

### PR DESCRIPTION
**Pull request by johnb432**
**When merged this pull request will:**
- Code cleanup.
- Moved event `"ace_cargoLoaded"` from being called within event `"ace_loadCargo"` to being called in `FUNC(loadItem)`.
- Added icon to renaming action (icon is from base game).
- Deleted `FUNC(handleDeleted)`, as it shared close to the entire code with `FUNC(handleDestroyed)`.
- `FUNC(handleDestroyed)` is now triggered when an entity is killed or deleted. It deletes all loaded cargo items of a vehicle. It also checks if item was in a vehicle as cargo and removes itself from said vehicle's cargo.
- Deleted `FUNC(validateCargoSpace)`, as it serves no purpose.
- Added config case verification to various functions. This should remove any case sensitivity.
- Changed a couple of condition checking instances, such as:
```sqf
private _hasCargoConfig = 1 == getNumber (configOf _x >> QGVAR(hasCargo));
private _hasCargoPublic = _x getVariable [QGVAR(hasCargo), false];
(_hasCargoConfig || {_hasCargoPublic})
```
is now
```sqf
_x getVariable [QGVAR(hasCargo), getNumber (configOf _x >> QGVAR(hasCargo)) == 1]
```
They aren't quite the same, but I feel the updated version is what is intended.
- `FUNC(canUnloadItem)` received more conditions to check.
- Fixes #9109 by having `FUNC(initObject)` be executed every time as global and JIP when its size is changed.
- Applied the same fix to `FUNC(initVehicle)`, as it also had the same problem.
- The variable set on vehicles `QGVAR(space)` is kept up to date more rigorously and accurately. The integration of negative sizes of items has been improved.
- The cargo menu has been enhanced:
  - When an item is not unloadable (= its size is negative), the item will be displayed in red in the menu,
  - Added tooltips showing the cargo size of an object or if it's not unloadable.
  - Does more checks for invalid states (e.g. if the vehicle has been locked or if there is any cargo left).
- Some files in `common` were also cleaned up, as they were used.
- In `dragging`, `FUNC(dropObject_carry)` was improved so that it no longer drops the carried item if it's possible to load the carried item.
- Some settings files outside of `cargo` had their categories changed, as localized vs. unlocalized entries, despite resulting in the same display name, are not considered to be the same category in CBA settings. If the categories in `cargo` had only been changed, there would have been two `ACE Logistics` categories in the CBA settings.
- Improved documentation.

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

**Comment by rautamiekka** [comment]
Damn, that's a lot.

There are so many simple checks behind lazy evals, many even in succession, that will surely do more harm than good, like the `(_hasCargoConfig || {_hasCargoPublic})` in the OP or booleans similar to `alive ACE_player`, though.

File: addons/cargo/functions/fnc_unloadItem.sqf [hunk 1]
```diff
@@ -1,57 +1,74 @@
...
+
 if (_object isEqualType objNull) then {
     detach _object;
+
     // hideObjectGlobal must be executed before setPos to ensure light objects are rendered correctly
-    // do both on server to ensure they are executed in the correct order
+    // Do both on server to ensure they are executed in the correct order
     [QGVAR(serverUnload), [_object, _emptyPosAGL]] call CBA_fnc_serverEvent;
 } else {
     _object = createVehicle [_item, _emptyPosAGL, [], 0, "NONE"];
```
Comment (LinkIsGrim) [inline addons/cargo/functions/fnc_unloadItem.sqf:74]: Not the point of the PR, but unloading always being handled on the server makes more sense IMO, instead of this.

File: addons/cargo/functions/fnc_onMenuOpen.sqf [hunk 2]
```diff
@@ -22,37 +22,64 @@ params ["_display"];
...
+            _ctrl lbSetTooltip [_index, LLSTRING(unloadingImpossible)];
+            _ctrl lbSetColor [_index, [1, 0, 0, 1]]; // set text to red
+            _ctrl lbSetSelectColor [_index, [1, 0, 0, 1]];
+        };
+    } forEach (_vehicle getVariable [QGVAR(loaded), []]);
 
-    _label ctrlSetText format[localize LSTRING(labelSpace), [GVAR(interactionVehicle)] call DFUNC(getCargoSpaceLeft)];
-}, 0, []] call CBA_fnc_addPerFrameHandler;
+    _label ctrlSetText format [LLSTRING(labelSpace), (_vehicle call FUNC(getCargoSpaceLeft)) max 0];
+}, 0.25, GVAR(interactionVehicle)] call CBA_fnc_addPerFrameHandler;
```
Comment (LinkIsGrim) [inline addons/cargo/functions/fnc_onMenuOpen.sqf:85]: Maybe move PFH to separate function?

Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:85]: Also, don't like the increased delay. Could get iffy with multiple people unloading from same vehicle.

File: addons/cargo/functions/fnc_onMenuOpen.sqf [hunk 3]
```diff
@@ -22,37 +22,64 @@ params ["_display"];
...
+    if (isNil "_display") exitWith {
+        _pfhID call CBA_fnc_removePerFrameHandler;
     };
 
-    if (isNull GVAR(interactionVehicle) || {(([ACE_player, GVAR(interactionVehicle)] call EFUNC(interaction,getInteractionDistance)) >= MAX_LOAD_DISTANCE) && {(vehicle ACE_player) != GVAR(interactionVehicle)}}) exitWith {
+    // Close menu if in invalid state
+    if (
+        !alive _vehicle ||
+        {locked _vehicle >= 2} ||
+        {!(_vehicle getVariable [QGVAR(hasCargo), getNumber (configOf _vehicle >> QGVAR(hasCargo)) == 1])} ||
```
Comment (LinkIsGrim) [inline addons/cargo/functions/fnc_onMenuOpen.sqf:43]: Possible to check against `vehicle getVariable [QGVAR(loaded), []]` instead here to avoid the config lookup? Though I don't see the need to close the menu if Cargo is empty.

File: addons/cargo/functions/fnc_onMenuOpen.sqf (hunk 2 above)
Comment (johnb432) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:85]: > Maybe move PFH to separate function?

It feels unnecessary, as the nearly the entire function is the definition of the PFH.

> Also, don't like the increased delay. Could get iffy with multiple people unloading from same vehicle.

Fair, I just wanted to avoid looking up info, such as size and cargo space, every frame.

File: addons/cargo/functions/fnc_onMenuOpen.sqf (hunk 3 above)
Comment (johnb432) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:43]: > Possible to check against `vehicle getVariable [QGVAR(loaded), []]` instead here to avoid the config lookup? Though I don't see the need to close the menu if Cargo is empty.

I don't think it's possible to use `QGVAR(loaded)` instead, as they don't serve quite the same purpose. It would mean you can't open the cargo menu if there are no items loaded.

We could cache all of the config entries in hashmaps.

Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:43]: Remind me what `QGVAR(hasCargo)` does at all then, because it seems redundant

Comment (johnb432) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:43]: It allows the cargo of a vehicle to be enabled/disabled.

Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:43]: Set that variable during initVehicle/initObject instead then?

Comment (johnb432) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:43]: https://github.com/acemod/ACE3/pull/9617/files#diff-ebaa31fadbcbd31a401d08ae98225fd53bc7a290710ce2cd6899e4311be561b4R41 worries me a bit - I'm honestly not quite sure if the comments means that the variable is API or not. Either way, it can be changed mid mission using `FUNC(setSpace)`.

File: addons/cargo/functions/fnc_onMenuOpen.sqf (hunk 2 above)
Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_onMenuOpen.sqf]: Could be solved by updating menu on unload/load global event instead but it's more refactoring

File: addons/cargo/functions/fnc_onMenuOpen.sqf (hunk 3 above)
Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:43]: Not public. Should only be touched by `FUNC(setSpace)` (which if set to 0 disables cargo and thus should disable the menu). Go wild.

File: addons/cargo/functions/fnc_onMenuOpen.sqf (hunk 2 above)
Comment (johnb432) [reply addons/cargo/functions/fnc_onMenuOpen.sqf]: I don't think it's worth it, at least for now.

File: addons/cargo/functions/fnc_onMenuOpen.sqf (hunk 3 above)
Comment (johnb432) [reply addons/cargo/functions/fnc_onMenuOpen.sqf]: It no longer checks for the config entry, as it assumes the condition to open the menu already did.

File: addons/cargo/functions/fnc_unloadItem.sqf (hunk 1 above)
Comment (johnb432) [reply addons/cargo/functions/fnc_unloadItem.sqf:74]: If you want to add a comment, what would it be? I'm not sure if such a comment is necessary.

Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_unloadItem.sqf:74]: I'm referring to line 74 and below, virtual items being created locally and then unloaded vs existing items being unloaded on the server and then passed along

**Comment by LinkIsGrim** [comment]
GVAR(interactionVehicle) and GVAR(interactionParadrop) variables keep their values after the menu is closed, this can be problematic.

Found a bug when cancelling item load but trying to get consistent repro atm.

**Comment by LinkIsGrim** [comment]
Carry item -> Load with LMB -> Cancel load: Item remains attached, carrying hint is gone

**Comment by johnb432** [comment]
> GVAR(interactionVehicle) and GVAR(interactionParadrop) variables keep their values after the menu is closed, this can be problematic.

In what way? For `GVAR(interactionVehicle)` I can understand, but not for `GVAR(interactionParadrop)`.

File: addons/cargo/functions/fnc_onMenuOpen.sqf [hunk 4]
```diff
@@ -22,37 +22,67 @@ params ["_display"];
...
+
     disableSerialization;
+
     private _display = uiNamespace getVariable QGVAR(menuDisplay);
-    if (isnil "_display") exitWith {
-        [_this select 1] call CBA_fnc_removePerFrameHandler;
+
+    if (isNil "_display") exitWith {
+        GVAR(interactionVehicle) = nil;
+        GVAR(interactionParadrop) = nil;
```
Comment (johnb432) [inline addons/cargo/functions/fnc_onMenuOpen.sqf:37]: This won't work, as `FUNC(startUnload)` requires both variables.

Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_onMenuOpen.sqf:37]: Black magic is making it work regardless. Spooky.

File: addons/common/define.hpp
Comment (LinkIsGrim) [inline addons/common/define.hpp:1]: This really should've been done in a separate PR

**Review by LinkIsGrim — APPROVED** [review]

File: addons/cargo/functions/fnc_unloadItem.sqf (hunk 1 above)
Comment (johnb432) [reply addons/cargo/functions/fnc_unloadItem.sqf]: So you would want the server to handle it fully? So decide if it needs to be actually unloaded or created?

Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_unloadItem.sqf]: Yes, would also make it easier to make everything event-based for the menu instead of running a PFH

**Comment by johnb432** [comment]
https://github.com/acemod/ACE3/pull/9493#pullrequestreview-1726040535 is still relevant. Unfortunately no text is displayed, so this issue should be resolved before pushed.

File: addons/cargo/functions/fnc_unloadItem.sqf (hunk 1 above)
Comment (johnb432) [reply addons/cargo/functions/fnc_unloadItem.sqf]: It's not easily possible, given the current structure, seeing as we need the reference to the object after it's been created.

I guess we could split `FUNC(unloadItem)` and `FUNC(paradropItem)` in half, so it would result in `FUNC(unloadItem)`, `FUNC(paradropItem)` and `FUNC(paradropItemServer)`, where `FUNC(unloadItem)` and FUNC(paradropItem) would remove the server-sided code.

Comment (LinkIsGrim) [reply addons/cargo/functions/fnc_unloadItem.sqf]: IMO should be done in a separate PR to limit scope

Comment (johnb432) [reply addons/cargo/functions/fnc_unloadItem.sqf]: > IMO should be done in a separate PR to limit scope

That's just what I was thinking.

**Comment by johnb432** [comment]
> [#9493 (review)](https://github.com/acemod/ACE3/pull/9493#pullrequestreview-1726040535) is still relevant. Unfortunately no text is displayed, so this issue should be resolved before being merged.

I'm getting an error code `5`. It happens randomly when you repeatably try to paradrop an item.

My guess is that the cargo menu doesn't close fast enough and `EFUNC(common,progressBar)` does not like that. I'll try delaying it by a frame.

File: addons/dragging/functions/fnc_dropObject_carry.sqf [hunk 5]
```diff
@@ -30,8 +30,17 @@ private _inBuilding = _unit call FUNC(isObjectOnObject);
 [QEGVAR(common,fixCollision), _unit] call CBA_fnc_localEvent;
 [QEGVAR(common,fixCollision), _target, _target] call CBA_fnc_targetEvent;
 
-// Release object
-detach _target;
+getCursorObjectParams params ["_cursorObject", "", "_distance"];
```
Comment (johnb432) [inline addons/dragging/functions/fnc_dropObject_carry.sqf:33]: I don't know how I missed this, but you can load items into a vehicle, when your back is turned to it.
`getCursorObjectParams` acts like `cursorTarget`, meaning it's very broad in its selection. It means also that you can't place a cargo item next to a vehicle, cause it will automatically give you the load option.

I was thinking of switching to the `cursorObject` command, but using `getCursorObjectParams select 2` for the distance. This would allow for finer control.

Thoughts?

Comment (LinkIsGrim) [reply addons/dragging/functions/fnc_dropObject_carry.sqf:33]: At that point why not just do `ACE_player distance cursorObject` ?

Comment (johnb432) [reply addons/dragging/functions/fnc_dropObject_carry.sqf:33]: > At that point why not just do `ACE_player distance cursorObject` ?

Could do - but do you agree with the change in principle?

Comment (LinkIsGrim) [reply addons/dragging/functions/fnc_dropObject_carry.sqf:33]: Yes

Comment (johnb432) [reply addons/dragging/functions/fnc_dropObject_carry.sqf]: Implemented - can you review the most recent changes again?

**Review by LinkIsGrim — APPROVED** [review]
